capstone/fitness/src/main/java/com/google/sps/servlets/UpdateProgressServlet.java

```diff
@@ -0,0 +1,144 @@
+import java.io.IOException;
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Date;
+import java.util.List;
+import javax.servlet.RequestDispatcher;
+import javax.servlet.ServletException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/update-progress")
+public class UpdateProgressServlet extends HttpServlet {
+
+  static Gson gson;  
+  
+  @Override
+  public void init() {
+    gson = new Gson();
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
+    // Display the complete session screen which then sends a POST to this server.
+    RequestDispatcher view = request.getRequestDispatcher("complete-session.html");
+    view.forward(request, response);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
+    
+    // Get the users email.
+    UserService userService = UserServiceFactory.getUserService();
+    String userEmail = userService.getCurrentUser().getEmail();
+
+    // Redirect if user not found.
+    if(userEmail == null) {
+      RequestDispatcher view = request.getRequestDispatcher("/");
+      view.forward(request, response);
+      System.out.println("user is null");
+    }
+
+    // Set up datastore and get query.
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query = new Query(CreateProfileServlet.USER_ENTITY);
+    PreparedQuery results = datastore.prepare(query);
+
+    // Get user from datastore
+    // I tried some other methods for getting the user but they were giving me problems.
+    // This approach is O(n) which should be the same as the built in methods.
+    Entity user=null;
+    loopThroughComments: 
+    for(Entity e : results.asIterable()) {
+        String key = e.getKey().toString();
+        if(!(key == null) && equalKeys(userEmail, key)) {
+          user = e;
+        }
+    }
+
+    // Redirect if user not found.
+    if(user == null) {
+      RequestDispatcher view = request.getRequestDispatcher("/");
+      view.forward(request, response);
+      System.out.println("user is null");
+    }
+
+    // Get the users current progress string (JSON format).
+    String progressJson = (String) (user.getProperty(CreateProfileServlet.PROGRESS_PROPERTY));
+
+    // Get the users marathon length
+    double m = (Double) user.getProperty(CreateProfileServlet.MARATHON_LENGTH_PROPERTY);
+    Float marathonLength = (float) m;
+
+    // Should never happen if user is not null.
+    if(progressJson == null) {
+      progressJson = "[]";
+    }
+
+    // Get parameters from progress update and calculate total time.
+    Float hours = Float.parseFloat(request.getParameter("hours"));
```

Comment:
why use float instead of int here?

Reply:
To calculate total time, I have to divide minutes by 60 and seconds by 3600, so I thought I should just declare it as a float. It will also be stored as a float.